### PR TITLE
[DO NOT MERGE YET] revamp node.js build, package, and build process; add semistandard li…

### LINKED ISSUE #284: Output build log of node module on stdout
I think it'd be more convenient to have the build log visible while building, as opposed to having to go digging for `build_log` every time it fails. What do you think?

### LINKED ISSUE #285: Move package.json to repository root
This would allow one to put `uws` as a [git dependency](https://docs.npmjs.com/files/package.json#git-urls-as-dependencies) in their `package.json`.

### PR DESCRIPTION
**Pull request by SEAPUNK**
…nter
### DO NOT MERGE JUST YET. Because of the size of this change, I want people to test this PR first. I published a package for testing: `npm install uws-new-build`

---

TODO before merging:
- [x] Primus integration tests
- [ ] engine.io integration tests
- [ ] Discuss cross-platform binary building and packaging (Windows, Darwin binaries)
- [ ] Fix typo (see commit comment)
- [ ] Make sure new build process will work
- [ ] Get god-emperor @alexhultman's approval

---

Fixes #285
Fixes #284

This completely changes how node.js is packaged, and built.
This also adds the semistandard linter.
- .npmignore added (this prevents files from getting into the published npm package, a gitignore for npm)
- package.json moved to project root (allows for installing from git)
- Replaces platform-specific packaging and build Makefiles with
  cross-platform node.js scripts
- Added semistandard linter; npm test now lints the .js source
- Reorganized some miscellaneous node.js scripts to nodejs/misc/
- Splits the native module requirer into lib/native.js, for the build
  scripts (prevents build false-negatives in case the JS code has errors),
  and people who want to load just the native module (e.g. Primus)

---

For you, @alexhultman, this will change how you package and publish the module. This is what you need to do before publishing:

_at the project root_
- `npm install` (installs the development dependencies for testing)
- `npm test` (tests the code, which as of right now, is just the `semistandard` linter)
- `npm publish`. The npm `prepublish` script automagically builds the (Linux) node.js binary addons for you, so you don't have to do any `make`ing beforehand. Darwin and Windows, we need to figure something out for that.

### CODE REVIEW

**Comment by SEAPUNK** [comment]
I'm testing Primus integration right now.

Also, I just noticed that your `uws` package provides prebuilt binaries for Darwin, and Windows. How do you usually do that?

**Comment by SEAPUNK** [comment]
Primus tests look fine, aside from #299, so that's good.

**Comment by SEAPUNK** [comment]
@alexhultman Now that I think of it, because the build process is done in scripts/build.js, we could even remove node-gyp entirely, and use the raw compilation executables, with their flags.

I'm not entirely sure _how_ possible it is, but we could always have the script to fallback to node-gyp if everything else fails, as a sort-of last resort.

File: nodejs/binding.gyp [hunk 1]
```diff
@@ -29,7 +29,7 @@
             'CLANG_CXX_LIBRARY': 'libc++',
             'GCC_GENERATE_DEBUGGING_SYMBOLS': 'NO',
             'GCC_ENABLE_CPP_EXCEPTIONS': 'YES',
-            'GCC_THREADSAFE_STATICS': 'YES',
+            'GCC_THREADSAFE_STATICnoS': 'YES',
```
Comment (SEAPUNK) [inline nodejs/binding.gyp]: I need to fix this typo.

**Comment by silverwind** [comment]
Looks like some nice work. These changes will also make it easier to contribute for Node.js folks.

**Comment by SEAPUNK** [comment]
@silverwind Does the new build of uws work on your droppy platform?

**Comment by SEAPUNK** [comment]
I'm going to rebase and try out a new compilation method sometime later.

**Comment by silverwind** [comment]
Can conform this works after adding `git://github.com/SEAPUNK/uWebSockets.git#new-nodejs-build` to my `package.json`and reinstalling. 👍 from me.

**Comment by SEAPUNK** [comment]
@alexhultman When will be a good time to talk about this PR on Gitter? I have a few questions I want to get out of the way.

File: .npmignore [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+# .npmignore specific
+
+nodejs/tests/
+nodejs/misc/
+
+autobahn/
+benchmarks/
+examples/
+ssl/
+tests/
```
Comment (SEAPUNK) [inline .npmignore]: I'll look into it.

**Comment by silverwind** [comment]
@alexhultman any comments?

**Comment by SEAPUNK** [comment]
I told Alex that I'd be redoing parts of this later, but I haven't gotten around to it yet.

**Comment by SEAPUNK** [comment]
Okay, I think I know exactly what I want to do with this regarding build and publish process. I'll play with this PR and CI integration throughout this week.

Ref: https://github.com/SEAPUNK/uWebSockets/issues/1

**Comment by SEAPUNK** [comment]
I'll close this for now -- it might be a while before this is ready.

**Pull request closed without merging**